src/qt/netwatch.cpp

```diff
@@ -0,0 +1,829 @@
+    }
+    return false;
+}
+
+void LogEntry::makeWeak()
+{
+    const meta_t n = *(meta_t*)data;
+    if ((n >> 30) != 2) {
+        return;
+    }
+    CTransactionRef * const ptx_old = get<CTransactionRef>();
+    CTransactionRef tx = *ptx_old;  // save a copy
+    ptx_old->~CTransactionRef();
+    CTransactionWeakref * const ptx_new = get<CTransactionWeakref>();
+    new (ptx_new) CTransactionWeakref(tx);
+    *((meta_t*)data) |= (3 << 30);
+}
+
+class NetWatchValidationInterface final : public CValidationInterface {
+private:
+    NetWatchLogModel& model;
+
+public:
+    NetWatchValidationInterface(NetWatchLogModel& model_in) : model(model_in) {}
+    void ValidationInterfaceUnregistering();
+
+    void UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload);
+    void TransactionAddedToMempool(const CTransactionRef &);
+};
+
+void NetWatchValidationInterface::ValidationInterfaceUnregistering()
+{
+    model.OrphanedValidationInterface();
+}
+
+void NetWatchValidationInterface::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload)
+{
+    if (pindexNew != pindexFork) {
+        model.LogBlock(pindexNew);
+    }
+}
+
+void NetWatchValidationInterface::TransactionAddedToMempool(const CTransactionRef &ptx)
+{
+    model.LogTransaction(ptx);
+}
+
+NetWatchLogModel::NetWatchLogModel(QWidget *parent) :
+    QAbstractTableModel(parent),
+    widget(parent),
+    validation_interface(new NetWatchValidationInterface(*this))
+{
+    RegisterValidationInterface(validation_interface);
+}
+
+NetWatchLogModel::~NetWatchLogModel()
+{
+    LOCK(cs);
+    if (validation_interface) {
+        UnregisterValidationInterface(validation_interface);
+        delete validation_interface;
+        validation_interface = nullptr;
+    }
+}
+
+void NetWatchLogModel::OrphanedValidationInterface()
+{
+    LOCK(cs);
+    delete validation_interface;
+    validation_interface = nullptr;
+}
+
+int NetWatchLogModel::rowCount(const QModelIndex& parent) const
+{
+    LOCK(cs);
+    return log.size() - logskip;
+}
+
+int NetWatchLogModel::columnCount(const QModelIndex& parent) const
+{
+    return NWLMHeaderCount;
+}
+
+QVariant NetWatchLogModel::data(const CBlockIndex& blockindex, int txout_index, const Header header) const
+{
+    switch (header) {
+        case NWLMH_TYPE:
+            return LogEntry::LogEntryTypeAbbreviation(LogEntry::LET_BLOCK);
+        case NWLMH_ID:
+            return QString::fromStdString(blockindex.GetBlockHash().GetHex());
+        case NWLMH_ADDR:
+        case NWLMH_VALUE: {
+            if (blockindex.nTx == 0 || !(blockindex.nStatus & BLOCK_HAVE_DATA)) {
+                return QVariant();
+            }
+            CBlock block;
+            if (!ReadBlockFromDisk(block, &blockindex, Params().GetConsensus())) {
+                // Indicate error somehow?
+                return QVariant();
+            }
+            assert(block.vtx.size());
```

Comment:
Nit: `!block.vtx.empty()` communicates intention more clearly than `block.vtx.size()`?

Reply:
Not sure. The intent is only to make sure `vtx[0]` below is an acceptable dereference.

The actual check for block data is above, with `if (blockindex.nTx == 0 || !(blockindex.nStatus & BLOCK_HAVE_DATA))`

Reply:
I see -- makes sense :-)